docs/docs/internals/explicit-nulls.md

```diff
@@ -0,0 +1,100 @@
+---
+layout: doc-page
+title: "Explicit Nulls"
+---
+
+The "explicit nulls" feature (enabled via a flag) changes the Scala type hierarchy
+so that reference types (e.g. `String`) are non-nullable. We can still express nullability
+with union types: e.g. `val x: String|Null = null`.
+
+The implementation of the feature in dotty can be conceptually divided in several parts:
+  1. changes to the type hierarchy so that `Null` is only a subtype of `Any`
+  2. a "translation layer" for Java interop that exposes the nullability in Java APIs
+  3. a "magic" `JavaNull` type (an alias for `Null`) that is recognized by the compiler and
+     allows unsound member selections (trading soundness for usability)
+
+Feature Flag
+------------
+Explicit nulls are disabled by default. They can be enabled via `-Yexplicit-nulls` defined in
+`ScalaSettings.scala`. All of the explicit-nulls-related changes should be gated behind the flag.
+
+Type Hierarchy
+--------------
+We change the type hierarchy so that `Null` is only a subtype of `Any` by:
+  - modifying the notion of what is a nullable class (`isNullableClass`) in `SymDenotations`
+    to include _only_ `Null` and `Any`
+  - changing the parent of `Null` in `Definitions` to point to `Any` and not `AnyRef`
+  - changing `isBottomType` and `isBottomClass` in `Definitions`
+
+Java Interop
+------------
+TODO(abeln): add support for recognizing nullability annotations a la
+https://kotlinlang.org/docs/reference/java-interop.html#nullability-annotations
+
+The problem we're trying to solve here is: if we see a Java method `String foo(String)`,
+what should that method look like to Scala?
+  - since we should be able to pass `null` into Java methods, the argument type should be `String|JavaNull`
+  - since Java methods might return `null`, the return type should be `String|JavaNull`
+
+`JavaNull` here is a type alias for `Null` with "magic" properties (see below).
+
+At a high-level:
+  - we track the loading of Java fields and methods as they're loaded by the compiler
+  - we do this in two places: `Namer` (for Java sources) and `ClassFileParser` (for bytecode)
+  - whenever we load a Java member, we "nullify" its argument and return types
+
+The nullification logic lives in `JavaNullInterop.scala`, a new file.
+
+The entry point is the function `def nullifyMember(sym: Symbol, tp: Type)(implicit ctx: Context): Type`
+which, given a symbol and its "regular" type, produces what the type of the symbol should be in the
+explicit nulls world.
+
+In order to nullify a member, we first pass it through a "whitelist" of symbols that need
+special handling (e.g. `constructors`, which never return `null`). If none of the "policies" in the
+whitelist apply, we then process the symbol with a `TypeMap` that implements the following nullification
```

Comment:
that implements the following nullification -> that implements the nullification function defined in XYZ

Reply:
remove function def below